envoy/buffer/buffer.h

```diff
@@ -458,6 +458,31 @@ class Instance {
     writeInt<ByteOrder::BigEndian, T, Size>(value);
   }
 
+  /**
+   * Copy multiple arguments to the buffer. In order to improve performance, all the arguments in
+   * each call will be stored in contiguous memory. So this method is mainly used when a large
+   * amount of fine-grained data needs to be written to the buffer. The type of arguments must be
+   * automatically converted to `absl::string_view`.
+   * @param args A set of arguments with variable length and type. The arguments must be guaranteed
+   * to be safely converted to `absl::string_view`.
+   * @return The total size of the data copied to the buffer.
+   */
+  template <class... Args> size_t addFragments(Args&&... args) {
+    size_t total_size_to_write = 0;
+    if constexpr (sizeof...(args) > 0) {
+      constexpr auto WriteHelper = [](uint8_t** dst, absl::string_view v) {
+        memcpy(*dst, v.data(), v.size()); // NOLINT(safe-memcpy)
+        *dst += v.size();
+      };
+
+      total_size_to_write = (absl::string_view(args).size() + ...);
+      auto* dst_memory_to_write = inlineReserve(total_size_to_write);
+      (WriteHelper(&dst_memory_to_write, absl::string_view(args)), ...);
```

Comment:
I would have considered this to have been a call to a AddImpl which skips watermark checks instead of low level operations based on reservations.

Reply:
This is because we not only want to reduce watermark checks, but we also want to reduce some intermediate checks. **"One reservation and multiple writes no any more intermediate check"**. So `addImpl` is a bit too complicated for this.

From the benchmark results. Even compared to `OwnImpl` without any watermark check, now `addFragments` can get a good performance improvement.